### ISSUE TITLE
[css-text-3] Ogham Space Mark needs to disappear at the end of a line

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#4893]
http://unicode.org/reports/tr14/#BA

**Comment by fantasai** [comment]
Checked in this rule in https://github.com/w3c/csswg-drafts/commit/f047b93f7efb978664c902368013644006146f19

Agenda+ to verify with CSSWG.

**Comment by frivoal** [comment]
I agree with the general idea of this change, but it seems to me that while trailing OGHAM SPACE MARK with `white-space: normal | nowrap | pre-line` should get removed, I don't quite see why we'd drop those with `white-space:pre`.

As for pre-wrap and break-spaces, I am less sure. Probably do as we do with other kinds of spaces: keep it for `break-spaces`, and keep it + hang it for `pre-wrap`, but I am not 100% sure.

**Comment by fantasai** [comment]
@frivoal Oversight in the wording, you're right that only collapsible marks should be removed.

**Comment by fantasai** [comment]
@frivoal Feel free to adjust the wording if you've got a better suggestion. :)

**Comment by frivoal** [comment]
> Oversight in the wording, you're right that only collapsible marks should be removed.

After the edit, it now looks correct.

> As for pre-wrap and break-spaces, I am less sure.

Never mind, this is already covered, since the Ogham Space Mark is part of unicode category Zs, and thus of "other spaces separators", so that's handled already.

TL;DR: we're good.

**Comment by fantasai** [comment]
Chatted with @kojiishi , there's basically several ways we can handle this in the details. Note that from a practical perspective, it really does not matter which one we choose, since mixing OSM and spaces isn't a real use case, so we should do the most efficient thing for implementations.

- Include Ogham Space Mark among the white space characters trimmed at the end of a line ([phase 2](https://drafts.csswg.org/css-text-3/#white-space-phase-2)), but don't do anything with them when collapsing within the line ([phase 1](https://drafts.csswg.org/css-text-3/#white-space-phase-1)).
- Trim Ogham Space Mark as a separate step after trimming white space characters at the end of a line.
- Add Ogham Space Mark to the list of white space characters and collapse them just like spaces everywhere. (This is probably overkill, it makes the CSS document white space set different from CSS and HTML syntactic white space, and it likely has bad perf implications.)

The difference between 1 & 2 is what happens in the sequence `[Letter][Space][OSM][Space]`. 1 would trim everything after the Letter. 2 would leave the inner space intact.

**Comment by fantasai** [comment]
@kojiishi Also suggests marking this behavior at-risk for L3; I agree with that.

**Comment by fantasai** [comment]
@jfkthame @kojiishi Any thoughts on which option is easiest to implement? See https://github.com/w3c/csswg-drafts/issues/4893#issuecomment-613175286

**Comment by jfkthame** [comment]
In general, I'm guessing option 1 (include Ogham Space Mark among the white space characters trimmed at the end of a line) would be the most straightforward.

However, I have one somewhat-related question. The current text in 4.1.2 step 3 says:

>  A sequence at the end of a line of collapsible spaces is removed, and any trailing U+1680 OGHAM SPACE MARK is also removed if it’s ‘white-space’ property is ‘normal’, ‘nowrap’, or ‘pre-line’. 

ISTM this will lead to a surprising user experience in editable content, in that when U+1680 is typed at the end of the current text, it doesn't appear (as this rule removes it); but then when another letter is typed after it, the intervening space mark suddenly pops into view.

Does this case need to be addressed somehow? I'm not sure exactly where/how, though...

**Comment by frivoal** [comment]
the user experience with editable content when white-space isn't pre-wrap (or break-spaces) is already quite weird anyway, so I am not too sure we should be worried about that in particular.
